src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -409,5 +559,118 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::KDF(const uint8_t * ikm, const size_t
+{
+    VerifyOrReturnError(fe_length_bytes > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(raw_sig != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(raw_sig_length == (2u * fe_length_bytes), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(out_asn1_sig != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(out_asn1_sig_length >= (raw_sig_length + kMax_ECDSA_X9Dot62_Asn1_Overhead), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Min overhead of SEQUENCE is 3 bytes
+    constexpr size_t kMinOverhead = 1u + 2u; // 1 byte tag, (1 byte length, 1 byte data) or 3 bytes length
+
+    // Write both R an S integers past the overhead, we will shift them back later if we only needed 2 size bytes.
+    uint8_t * cursor = &out_asn1_sig[kMinOverhead];
+    size_t remaining = out_asn1_sig_length - kMinOverhead;
+
+    size_t integers_length = 0;
+
+    // Write R (first `fe_length_bytes` block of raw signature)
+    size_t integer_length = EmitDerIntegerFromRaw(&raw_sig[0], fe_length_bytes, cursor, remaining);
+    VerifyOrReturnError(integer_length != 0, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(integer_length < remaining, CHIP_ERROR_INTERNAL);
+    remaining -= integer_length;
+    integers_length += integer_length;
+    cursor += integer_length;
+
+    // Write S (second `fe_length_bytes` block of raw signature)
+    integer_length = EmitDerIntegerFromRaw(&raw_sig[fe_length_bytes], fe_length_bytes, cursor, remaining);
+    VerifyOrReturnError(integer_length != 0, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(integer_length <= remaining, CHIP_ERROR_INTERNAL);
+    integers_length += integer_length;
+
+    // We only support outputs that would use 1 or 2 bytes of DER length after the SEQUENCE tag
+    VerifyOrReturnError(integers_length <= UINT8_MAX, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // We now know the length of both variable sized integers in the sequence, so we
+    // can write the tag and length.
+    BufferWriter writer(out_asn1_sig, static_cast<uint16_t>(out_asn1_sig_length));
+
+    // Put SEQUENCE tag
+    writer.Put(kSeqTag);
+
+    // Put the length over 1 or two bytes depending on case
+    if (integers_length > 127u)
+    {
+        writer.Put(static_cast<uint8_t>(0x80u | 1u)); // Length is extended length, over 1 subsequent byte
+        writer.Put(static_cast<uint8_t>(integers_length));
+    }
+    else
+    {
+        // Length is directly in the first byte with MSB clear if <= 127.
+        writer.Put(static_cast<uint8_t>(integers_length));
+    }
+
+    // Put the contents of the integers previously serialized in the buffer.
+    // The writer.Put is memmove-safe, so the shifting will happen from the read
+    // of the same buffer where the write is taking place.
+    writer.Put(&out_asn1_sig[kMinOverhead], integers_length);
+
+    size_t actually_written = 0;
+    if (writer.Fit(actually_written))
```

Comment:
I believe code may be easier to follow if the 'return no error' is at the end. Then this becomes:

```cpp
VerifyOrReturnError(writer.Fit(actually_written), CHIP_ERROR_BUFFER_TOO_SMALL);

out_asn1_sig_actual_length = actually_written;
return CHIP_NO_ERROR;
```